pyspecProcScripts/first_level/QESR_rescale.py

```diff
@@ -1,31 +1,48 @@
-# note that this will fail if you don't have "QESR Conversion" set in your pyspecdata file
-dint_conversion = pyspec_config.get_setting("QESR Conversion")
-if dint_conversion is not None:
-    dint_conversion = float(dint_conversion)
-
+class calib_info (object):
+    "class for caching calibration info"
+    def __init__(self):
+        self.current_calib_name = None
+        self.current_diam_name = None
+    def use_calibration(self, calibration_name):
+        if calibration_name != self.current_calib_name:
+            assert type(calibration_name) is str, "the calibration name must be specified as a string!"
+            assert len(calibration_name)>0, "you MUST use a calibration name and set the values `[calibration_name] q` and `[calibration_name] conversion` in your _pyspecdata file!"
+            self.current_calib_name = calibration_name
+            # note that this will fail if you don't have "[calibration_name] conversion" and "[calibration_name] q" set in your pyspecdata file
+            self.default_Q = float(pyspec_config.get_setting(f"{calibration_name} Q"))
+            self.dint_conversion = float(pyspec_config.get_setting(f"{calibration_name} Conversion"))
+    def use_diameter(self, diameter_name):
+        if diameter_name != self.current_diam_name:
+            assert type(diameter_name) is str, "the diameter name must be specified as a string!"
+            assert len(diameter_name)>0, "you MUST use a diameter name and set the values `[diameter_name] q` and `[diameter_name] conversion` in your _pyspecdata file!"
+            self.current_diam_name = diameter_name
+            self.d = float(pyspec_config.get_setting(diameter_name)
+calibcache = calib_info()
 ureg = UnitRegistry(
     system="mks", autoconvert_offset_to_baseunit=True, auto_reduce_dimensions=True
 )
 Q_ = ureg.Quantity
-def QESR_scalefactor(d):
+def QESR_scalefactor(d, calibration_name=None, diameter_name=None):
     "determine the scaling factor for QESR"
+    calibcache.use_calibration(calibration_name)
+    calibcache.use_diameter(diameter_name)
     # {{{ determine the signal denominator from the parameters of interest
     G_R = Q_(*d.get_prop("Gain"))
     C_t = Q_(*d.get_prop("ConvTime"))
     power = Q_(*d.get_prop("Power"))
     B_m = Q_(*d.get_prop("ModAmp"))
-    Q = Q_(default_Q, "dimensionless")  # hard set Q value
+    Q = Q_(calibcache.default_Q, "dimensionless")  # hard set Q value
+    d = Q_(calibcache.d, "mm")  # hard set Q value
     n_B = Q_(1, "dimensionless")  # calculate this
     S = Q_(0.5, "dimensionless")
     c = Q_(
         1, "dimensionless"
     )  # the first fraction on pg 2-17 -- essentially the conversion factor
-    signal_denom = G_R * C_t * sqrt(power) * B_m * n_B * S * (S + 1) * Q
-    signal_denom = signal_denom.to(Q_("G") * sqrt(Q_("W")) * Q_("s"))
+    signal_denom = G_R * C_t * sqrt(power) * B_m * n_B * S * (S + 1) * Q * d**2
+    signal_denom = signal_denom.to(Q_("G") * sqrt(Q_("W")) * Q_("s") * Q_("m")**2)
```

Comment:
as I noted [here](https://www.wrike.com/open.htm?id=1031155470) the actual values of the conversion factor will need to be multiplied by d^2 (with d in mm) to compensate for the fact that we are now including d here.  We should only make this change for the new-style (named) conversion factors, so that we can still use the old code.